lib/puppet/type/zfs_share.rb

```diff
@@ -0,0 +1,141 @@
+Puppet::Type.newtype(:zfs_share) do
+  @doc = "Manage zfs shares"
+
+  require 'puppet/parameter/boolean'
+
+  ensurable
+
+  newparam(:zfs_name) do
+    desc "Name of zfs file system"
+    isnamevar
+
+    validate do |value|
+      unless value =~ /(?!\/|\d)^\w.*/
+        raise ArgumentError , "%s is not a valid zfs path."
+      end
+    end
+  end
+
+  newparam(:zpool) do
+    desc "Name of the zfs parent"
+
+    validate do |value|
+      unless value =~ /(?!\/|\d)^\w.*/
+        raise ArgumentError , "%s is not a valid zpool."
+      end
+    end
+  end
+
+  newparam(:share_title) do
+    desc "Override the title of the share"
+
+    validate do |value|
+      if value =~ /\s/
+        raise ArgumentError , "#{value} is not a valid share title"
+      end
+    end
+  end
+
+  newparam(:allow_ip_read, :array_matching => :all) do
```

Comment:
So, I'm not necessarily opposed to changing this pattern, but at the same time you haven't offered concrete examples of what would be better. I would like to move this forward, so please be more specific in what you would prefer to see, or let's defer those changes to a later iteration.